ui-spacetimechart: implements various features

jacomyal opened the pull request:
The related features from ticket [#502](https://github.com/osrd-project/osrd-confidential/issues/502) are the following:
- Inverser les axes du graphe (échanger les axes X et Y)
- Ajouter un snap aux PRs du sillon à l’outil de mesure
- Pouvoir masquer le quadrillage
- Différencier l’espace occupé par les voies uniques sur toute la largeur du graphe
- Afficher les temps de fermeture de postes
- Disposer d’une image PNG de la vue courante
- Permettre d’accéder à différents thèmes de couleur (fond, axes, lignes…)
- Garder un sillon sélectionné dans la vue lorsqu’on se déplace dans le temps à l’aide de la molette

I kept quite atomic commits, to make them easier to review.

jacomyal committed:
ui-spacetimechart: allows swaping axis

Details:
- Renames getX and getY as getTimePixel and getSpacePixel
- Propagates through context new information about axis (timeAxis,
  spaceAxis and swapAxis)
- Updates all existing layers to handle the new swapAxis boolean option
- Adds a new story that showcases the new swapAxis option

jacomyal committed:
ui-spacetimechart: implements snapping

Details:
- Adds new helpers to help drawing aliased discs, with some basic unit
  tests
- Improves picking, so that each segment and each point of interest has
  a unique ID
- Keeps track of what exactly is hovered (not only the path, but also
  the exact point or segment)
- Adds new helpers to snap a point to the hovered item when relevant,
  with some basic unit tests
- Splits useMouse into useMouseTracking (that just keeps track of the
  mouse state) and useMouseInteractions, that takes hoveredItem as an
  input, and handles all event handlers
- Adds new enableSnapping option
- Updates measuring story to showcase that new snapping feature

jacomyal committed:
ui-spacetimechart: adds various options

Details:
- Moves various components used in measuring story to a new
  src/stories/components.tsx file
- Fixes getDataToPoint and getPointToData functions with swapAxis option
- Implements new options hideGrid and hidePathLabels
- Renames swap-axis story to options story, and adds more options in it

jacomyal committed:
ui-spacetimechart: adds custom rendering examples

Details:
- Refactors a bit src/stories structure
- Adds a new additional-data.stories.tsx, that showcases how to render
  mono-track sections or closed-doors

jacomyal committed:
ui-spacetimechart: adds PNG export

jacomyal committed:
ui-spacetimechart: enables custom theming

Details:
- Moves everything related to graduations, paths and captions rendering
  into the context, under the `theme` key
- The theme is spread through the SpaceTimeChartContext, and is
  overridable from the SpaceTimeChart props
- Adds a new custom-styles.stories.tsx example, that showcases how to
  override the theme

jacomyal committed:
ui-spacetimechart: allows path focus during scroll

Details:
- Adds new geometry utils, to detect if a path is on screen or not, and
  unit tests for these utils
- Adds SpaceTimeChartContextType within all handler events, for
  convenience
- Adds a new story where the mouse wheel is used to scroll on the time
  axis, and keep a selected path always visible while scrolling

Yohh commented on ui-spacetimechart/src/stories/options.stories.tsx:
```diff
@@ -0,0 +1,197 @@
+      }}
+    >
+      <button
+        onClick={() =>
+          captureCanvases().then((blob) => FileSaver.saveAs(blob, 'space-time-chart.png'))
+        }
```
is there a reason to use `FileSaver` instead of the canvas method `toDataURL()` ?

jacomyal replied:
Using `toDataURL` allows opening the image in a new tab, but it's not enough to download the image as a local file (which was what I was aiming at).

And historically, saving a local file from a browser is a bit tricky (using an hidden `a` tag in the DOM, and differently in some browsers). So, I used the most used JS lib to handle that, [file-saver](https://www.npmjs.com/package/file-saver).

Yohh replied:
ok that was what I expected to see (add an hidden `a`)
thanks

jacomyal replied:
If you prefer that we cut the dependency, it's possible. But it's quite a light one, and it relieves us from maintaining some tricky (short, but still tricky) code that handles something `file-saver` does better.

Yohh replied:
no it's fine to me, I was just asking for "personal culture"

jacomyal replied:
ok :)

Yohh commented on ui-spacetimechart/src/utils/canvas.ts:
```diff
@@ -75,11 +77,96 @@ export function drawAliasedLine(
+ *
+ * Here are some examples to make it clearer what the output should look like:
+ *
+ * getAliasedDiscShape(0);
+ * // [1]
+ * getAliasedDiscShape(1);
+ * // [0, 1, 0,
+ * //  1, 1, 1,
+ * //  0, 1, 0]
+ * getAliasedDiscShape(2);
+ * // [0, 0, 1, 0, 0,
+ * //  0, 1, 1, 1, 0,
+ * //  1, 1, 1, 1, 1,
+ * //  0, 1, 1, 1, 0,
+ * //  0, 0, 1, 0, 0]
```
```suggestion
 * getAliasedDiscShape(0);
 * [1]
 * getAliasedDiscShape(1);
 * [0, 1, 0,
 *  1, 1, 1,
 *  0, 1, 0]
 * getAliasedDiscShape(2);
 * [0, 0, 1, 0, 0,
 *  0, 1, 1, 1, 0,
 *  1, 1, 1, 1, 1,
 *  0, 1, 1, 1, 0,
 *  0, 0, 1, 0, 0]
```

Math-R commented on ui-spacetimechart/src/lib/types.ts:
```diff
@@ -126,6 +133,37 @@ export type MouseState = {
+  };
+  spaceGraduationsStyles: Record<
+    number,
+    { width: number; color: string; opacity?: number; dashArray?: number[] }
```
Will spaceGradutationsStyles and timeGraduationsStyles  Records always be similar ?
If so we can encapsulate those properties in some interface ?

jacomyal replied:
Good point. I don't see what could change - and if it does, we can handle it then.

jacomyal committed:
ui-spacetimechart: adapts from code review

The code review comments are readable in PR #68.

Details:
- Merges spaceGraduationsStyles and timeGraduationsStyles types
  declarations
- Handles theme validation
- Refactors new component Cross out of DataLabel component in stories
- Renames path as pathId in PickingElement type
- Fixes various code cosmetics issues, bad comments and bad exports

Math-R commented on ui-spacetimechart/src/utils/snapping.ts:
```diff
@@ -0,0 +1,58 @@
+ * This function takes a point and a segment, and returns the point of the segment that is the
+ * closest to the input point.
+ */
+export function getClosestPointOnSegment(
```
I see there is no tolerance setting in your snap function, but you are using disks to perform the snap. 
But in some cases the cursor will move from a point on the curve to the end of the segment. And we cannot therefore visualize the data of the curves between this point and the end of the segment.

Yohh replied:
the context:
[Screencast from 2024-08-29 15-47-10.webm](https://github.com/user-attachments/assets/5e13f7a0-ba55-4591-9623-090847a134b7)

Yohh replied:
[Screencast from 2024-08-29 15-47-10.webm](https://github.com/user-attachments/assets/c7cee236-85f6-4b7e-95cf-fbf224a0c859)

jacomyal replied:
I get your point. Indeed, when the mouse is too close to a point, the mouse will never snap to the curve. I'm not sure it's an issue, though, since points might be more important than curves (in which case, they must be easier to catch).

What do you think about this issue @thibautsailly?

thibautsailly replied:
The maximum zoom level displays 1 minute over 96px.
The snapping engages when the cursor is at 12~13px from a point, or 7,8 seconds at this zoom level.
I don't think missing contextual informations for this amount of time could be a problem, but I'll ask engineers on Monday if it's acceptable or not from their point of view.

I'm anticipating that they will be interested with snapping at path intersections as well as points.

clarani commented on ui-spacetimechart/src/hooks/useMouseInteractions.ts:
```diff
@@ -184,7 +125,5 @@ export function useMouse(
-  }, [position]);
-
-  return mouseState;
+  }, [position.x, position.y]);
```
Do you have a special reason to use these dependencies instead of `[position]` ?

jacomyal replied:
If a new `position` object with the same `x` and `y` values is received (with something like `const newPosition = { ...position };` for instance), the old `[position]` dependencies version triggers the effect, while the new `[position.x, position.y]` does not.

clarani commented on ui-spacetimechart/src/hooks/useMouseInteractions.ts:
```diff
@@ -173,7 +105,16 @@ export function useMouse(
+    if (onMouseMove) {
+      onMouseMove({
+        position,
+        isHover: isHover,
```
```suggestion
        isHover,
```

clarani commented on ui-spacetimechart/src/hooks/useMouseTracking.ts:
```diff
@@ -0,0 +1,73 @@
+  // Bind event handlers:
+  useEffect(() => {
+    if (!dom) return;
+    dom.addEventListener('mousedown', downHandler);
```
Why do you use the dom here and document below ?

jacomyal replied:
We only want to trigger the `"mousedown"` event when it's done on the `dom` element. But once the user starts dragging / panning, we want to keep interacting even when the mouse leaves the `dom` element. That's why, for `"mousemove"` and `"mouseup"`, we bind the events to `document` instead.

clarani commented on ui-spacetimechart/src/stories/utils.ts:
```diff
@@ -40,3 +40,28 @@ export function zoom(
 }
+
+/**
+ * This function helps to pretty-print a Date as a length:
```
```suggestion
 * This function helps to pretty-print a Date as a length.
```

clarani commented on ui-spacetimechart/src/stories/options.stories.tsx:
```diff
@@ -15,16 +16,18 @@ import {
 
 /**
- * This story aims at showcasing how to swap time and space axis in a SpaceTimeChart.
+ * This story aims at showcasing how to handle panning and zooming in a SpaceTimeChart.
```
This description is not correct I think

clarani commented on ui-spacetimechart/src/stories/components.tsx:
```diff
@@ -0,0 +1,112 @@
+/**
+ * This component draws a dashed line from p1 to p2, using SVG:
+ */
+export const Line: FC<{
```
No need to export it for the moment

clarani commented on ui-spacetimechart/src/stories/components.tsx:
```diff
@@ -0,0 +1,112 @@
+        top: -CROSS_SIZE / 2,
+        background: 'black',
+      }}
+    />
```
nit: you can even create a small component `Cross` with these 2 divs

clarani commented on ui-spacetimechart/src/stories/components.tsx:
```diff
@@ -0,0 +1,112 @@
+ * This component renders a data label, to help bring context to the SpaceTimeChart:
+ */
+const CROSS_SIZE = 11;
+export const DataLabel: FC<{
```
nit: no need to export DataLabel for the moment

clarani commented on ui-spacetimechart/src/stories/additional-data.stories.tsx:
```diff
@@ -0,0 +1,291 @@
+        timeScale={60000 / state.xZoomLevel}
+        xOffset={state.xOffset}
+        yOffset={state.yOffset}
+        onPan={({ initialPosition, position, isPanning }) => {
```
Maybe it would be nice to extract this function in a helper since it is used in many stories

jacomyal replied:
Agreed.

jacomyal replied:
It will be done later.

clarani commented on ui-spacetimechart/src/utils/png.ts:
```diff
@@ -0,0 +1,37 @@
+
+  // Create a new canvas, on which the different layers will be drawn:
+  const canvas = document.createElement('CANVAS') as HTMLCanvasElement;
+  canvas.setAttribute('width', width + '');
+  canvas.setAttribute('height', height + '');
```
Use real casts 
```suggestion
  canvas.setAttribute('width', String(width));
  canvas.setAttribute('height', String(height));
```

jacomyal replied:
I personnally don't agree with that one. `String`, as well as `Boolean` and `Number`, are very misleading in JavaScript, because they are class constructors, initially designed to handle mutable objects-like reference-based strings, booleans and numbers - while we almost always handle those values as immutable value-based strings, booleans and numbers. The fact that we can use them as casting functions is weird, in terms of design.

I fear that some JavaScript beginner, seeing `const w = String(width);` might think is it a typo and should be `const w = new String(width);` instead, while `const w = width + '';` is unique and enough commonly used. I have the same opinion on `const isEmpty = !!array.length;` and `const now = +(new Date());`.

Now, if you want to enforce the use of these types of casting, we can add the [`no-implicit-coercion` ESLint rule](https://eslint.org/docs/latest/rules/no-implicit-coercion).

Math-R replied:
You can also do ``` `${width}` ``` ?

jacomyal replied:
Yes, it does the exact same as `width + ''` I think.

clarani commented on ui-spacetimechart/src/components/TimeCaptions.tsx:
```diff
@@ -88,9 +62,9 @@ const TimeCaptions: FC = () => {
+      breakpoints.some((breakpoint, i) => {
         if (pixelsPerMinute < breakpoint) {
-          labelLevels = LABELS_PRIORITIES[i];
+          labelLevels = timeCaptionsPriorities[i];
```
Should we add a test here if the user doesn't provide a timeCpationsPriorities long enough ?

clarani commented on ui-spacetimechart/src/components/TimeCaptions.tsx:
```diff
@@ -112,16 +86,16 @@ const TimeCaptions: FC = () => {
       // Render caption top border:
-      ctx.strokeStyle = '#979797';
-      ctx.lineWidth = 2;
+      ctx.strokeStyle = timeGraduationsStyles[1].color;
+      ctx.lineWidth = timeGraduationsStyles[1].width;
```
What if `timeGraduationsStyles.length === 1` ? Should we throw an error or use the default theme ?

clarani commented on ui-spacetimechart/src/components/TimeGraduations.tsx:
```diff
@@ -1,63 +1,10 @@
-};
-
-const TimeGraduations: FC = () => {
+export const TimeGraduations: FC = () => {
```
Remove the export since you already default export it at the end of the file

clarani commented on ui-spacetimechart/src/components/PathLayer.tsx:
```diff
@@ -237,18 +245,20 @@ export const PathLayer: FC<PathLayerProps> = ({
       const w = measure.width + 2 * padding;
-      const h = FONT_SIZE + 2 * padding;
-      ctx.fillStyle = WHITE_75;
+      const h = measure.actualBoundingBoxAscent + measure.actualBoundingBoxDescent + 2 * padding;
```
Can you add a intermediate const here ? I'm not used to see `measure.actualBoundingBoxAscent` and `measure.actualBoundingBoxDescent`
```suggestion
      const fontSize = measure.actualBoundingBoxAscent + measure.actualBoundingBoxDescent;
      const h = fontSize + 2 * padding;
```

jacomyal replied:
(I agree, but it won't be named `fontSize`, since it does not compute a font size, but the actual text bounding box height)

clarani commented on ui-spacetimechart/src/components/PathLayer.tsx:
```diff
@@ -237,18 +245,20 @@ export const PathLayer: FC<PathLayerProps> = ({
       ctx.fillRect(dx - padding, dy - h + padding, w, h);
       ctx.fillStyle = labelColor;
+      ctx.globalAlpha = 1;
       ctx.fillText(label, dx, dy - FONT_SIZE / 4);
```
Shouldn't we use the fontSize computed above here ? (`measure.actualBoundingBoxAscent + measure.actualBoundingBoxDescent`)

jacomyal replied:
No, since the above is not a font size - but indeed, since a custom theme can override the actual font size, we should not use `FONT-SIZE`. I'll fix that :+1:

clarani commented on ui-spacetimechart/src/utils/geometry.ts:
```diff
@@ -0,0 +1,64 @@
+
+  const d1 = subtract(s1To, s1From);
+  const d2 = subtract(s2To, s2From);
+  const delta = subtract(s2From, s1From);
```
You can compute the delta after the if l17

clarani commented on ui-spacetimechart/src/utils/geometry.ts:
```diff
@@ -0,0 +1,64 @@
+ */
+export function intersectsSegments(s1From: Point, s1To: Point, s2From: Point, s2To: Point) {
+  const crossProduct = (a: Point, b: Point) => a.x * b.y - a.y * b.x;
+  const subtract = (a: Point, b: Point) => ({ x: a.x - b.x, y: a.y - b.y });
```
The function `getDiff` in `utils/vector.ts` does the same thing than subtract

clarani commented on ui-spacetimechart/src/components/PathLayer.tsx:
```diff
@@ -89,7 +89,6 @@ export type PathLayerProps = {
  * - The "picking" shape (to handle interactions)
  */
 export const PathLayer: FC<PathLayerProps> = ({
-  index,
```
Removing this will break Manchette, can you add a commit to fix it ?

clarani commented on ui-spacetimechart/src/stories/scroll-navigation.stories.tsx:
```diff
@@ -0,0 +1,225 @@
+                  const pathPositionAtTime = getSpaceAtTime(path, centerTime);
+                  const pathYAtCenter = getSpacePixel(pathPositionAtTime);
+
+                  newState.yOffset -= (pathYAtCenter - height / 2) / 4;
```
Why divide by 4 here ? Is it only to make the animation smoother ?

jacomyal replied:
Yes, exactly. I added a `frictionCoefficient` to make the code easier to understand.

clarani commented on ui-spacetimechart/src/lib/types.ts:
```diff
@@ -89,13 +91,17 @@ export type PointToData = (point: Point) => DataPoint;
 
+// PICKING SPECIFIC TYPES:
+export type PickingElement =
+  | { type: 'point'; path: string; point: Point }
+  | { type: 'segment'; path: string; from: Point; to: Point };
```
```suggestion
  | { type: 'point'; pathId: string; point: Point }
  | { type: 'segment'; pathId: string; from: Point; to: Point };
```

jacomyal committed:
ui-manchette: fixes spacetimechart usage

This commit fixes how SpaceTimeChart is used from the Manchette, after
PR #68 is merged.

Math-R reviewed (approved):
lgtm =)
